### ISSUE TITLE
Correct behaviour of binders `GenericModelBinder` handles

### ISSUE THREAD

**Issue by dougbu** [aspnet/Mvc#4181]
It's theoretically possible to instantiate (say) a `DictionaryModelBinder<string, string>` and insert it into `MvcOptions.ModelBinders`. The binders `GenericModelBinder` handles (chooses, instantiates and invokes) are all `public` types in the same `namespace` as `GenericModelBinder` itself.

Unfortunately `KeyValuePairModelBinder`, `CollectionModelBinder` and the `CollectionModelBinder` subclasses do not do the minimal checks necessary to confirm they apply before starting work. For example they do not double-check the `ModelBindingContext.ModelType`. Worse, most have `Debug` assertions for cases that can easily occur when used like any other `IModelBinder` implementation.

Suggest we pick one of the following:
1. Move these model binders to a `.Internal` namespace and add a few _more_ `Debug` assertions.
2. Flesh out these model binders so they can be instantiated and used directly.

----

Footnote: With the correct checks in place, could make this part of model binding more efficient: Change`GenericModelBinder` to place  `MvcOptions.ModelBinders` it instantiates earlier than itself in `MvcOptions.ModelBinders`. Then each binder needed would be instantiated just once, lazily.

**Comment by dougbu** [comment]
Partially addressed in d3c2463. `CollectionModelBinder` can still not be instantiated and added to `MvcOptions.ModelBinders` but is now a bit more efficient.

**Comment by Eilon** [comment]
What's the actual issue here? What can go wrong, and when?

**Comment by dougbu** [comment]
If a user instantiates any of these binders and adds it to `MvcOptions.ModelBinders`, that binder will always attempt to run to completion and will hit issues. Depending on the property type and particular binder, `Debug` assertions, NREs, and other errors are very likely. Bottom line these classes implement `IModelBinder` but can't be used like any other implementation of that `interface`.

**Comment by Eilon** [comment]
Ah ok that sounds like some good cleanup to do then.

**Comment by dougbu** [comment]
This issue is not relevant now that fb81a5e is in. As `IModelBinderProvider` split from `IModelBinder`, `GenericModelBinder` went away and providers such as `ArrayModelBinderProvider` and `CollectionModelBinderProvider` make users instantiating the binders themselves a non-scenario. Nothing left to do here.

**Comment by dougbu** [comment]
Marked invalid since fb81a5e rendered it so and this issue wasn't a duplicate of anything else.
